cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
   }
 
   public void upgradeVersionKeys() {
-    MDSKey startKey = new MDSKey.Builder().add(TYPE_APP_META).build();
-    Map<MDSKey, ApplicationMeta> appMetas = listKV(startKey, ApplicationMeta.class);
-    Map<MDSKey, ApplicationMeta> newAppMetas = new HashMap<>();
-    for (Map.Entry<MDSKey, ApplicationMeta> oldKey : appMetas.entrySet()) {
-      MDSKey newKey = appendDefaultVersion(oldKey.getKey());
-      newAppMetas.put(newKey, oldKey.getValue());
+    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_STARTED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_SUSPENDED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, RunRecordMeta.class);
+  }
+
+  public <T> void upgradeVersionKeys(String recordType, Type typeOfT) {
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
+    Map<MDSKey, T> oldMap = listKV(startKey, typeOfT);
+    Map<MDSKey, T> newMap = new HashMap<>();
+    for (Map.Entry<MDSKey, T> oldEntry : oldMap.entrySet()) {
+      MDSKey newKey = appendDefaultVersion(recordType, typeOfT, oldEntry.getKey());
+      newMap.put(newKey, oldEntry.getValue());
     }
 
     //Delete old rows
     deleteAll(startKey);
```

Comment:
if we delete all keys and something goes wrong after this we will not be able to recover.